File: Day4/solution.js
const fs = require('fs');

const data = fs
    .readFileSync('./input.txt', 'utf-8')
    .split('\r\n\r\n')
    .map((e) => {
        return e.replace(/\r\n/g, ' ');
    })
    .map((line) => generatePassportObject(line));

function generatePassportObject(string) {
    let obj = {};

    string.split(' ').map((e) => {
        let [key, value] = e.split(':');
        obj[key] = value;
    });

    return obj;
}

const part1 = data.filter((passport) => {
    let keys = Object.keys(passport);

    let isValid = true;

    keys.includes('byr') ? '' : (isValid = false);
    keys.includes('iyr') ? '' : (isValid = false);
    keys.includes('eyr') ? '' : (isValid = false);
    keys.includes('hgt') ? '' : (isValid = false);
    keys.includes('hcl') ? '' : (isValid = false);
    keys.includes('ecl') ? '' : (isValid = false);
    keys.includes('pid') ? '' : (isValid = false);

    if (isValid) {
        return passport;
    }
});

console.log('Part 1: ', part1.length);

const part2 = part1
    .filter((passport) => {
        let { eyr, iyr, byr } = passport;

        eyr = parseInt(eyr);
        iyr = parseInt(iyr);
        byr = parseInt(byr);

        let valid = true;

        if (eyr === NaN || eyr < 2020 || eyr > 2030) {
            valid = false;
        }
        if (iyr === NaN || iyr < 2010 || iyr > 2020) {
            valid = false;
        }
        if (byr === NaN || byr < 1920 || byr > 2002) {
            valid = false;
        }

        if (valid) {
            return passport;
        }
    })
    .filter((passport) => {
        let hgt = parseInt(passport.hgt.match(/\d+/));

        if (passport.hgt.match(/cm/)) {
            if (hgt < 150 || hgt > 193) {
                return false;
            }
        } else if (passport.hgt.match(/in/)) {
            if (hgt < 59 || hgt > 76) {
                return false;
            }
        } else {
            return false;
        }

        return passport;
    })
    .filter((passport) => {
        if (passport.hcl.match(/#[a-f0-9]{6}/)) {
            return passport;
        } else {
            return false;
        }
    })
    .filter((passport) => {
        const eyeColors = ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'];
        if (eyeColors.includes(passport.ecl)) {
            return passport;
        } else {
            return false;
        }
    })
    .filter((passport) => {
        if (passport.pid.length === 9 && passport.pid.match(/\d{9}/)) {
            return passport;
        } else {
            return false;
        }
    });

console.log('Part 2: ', part2.length);

File: Day4/solution.py
import re

passports = [x.replace('\n', ' ')
             for x in open('./input.txt').read().split('\n\n')]

required_fields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']

valid_passport_count = 0
valid_passport_count_part2 = 0
for passport in passports:
    # Part 1
    fields = {}
    for field in re.findall('(.{3}):', passport):
        fields[field] = 1

    valid = True
    for req_field in required_fields:
        if req_field not in fields:
            valid = False
            break

    if valid:
        valid_passport_count += 1

        # Part 2
        for req_field in required_fields:
            text = re.search(
                rf'{req_field}:[\w\d#]+', passport).group().split(':')[1]

            if req_field == 'byr':
                if not 1920 <= int(text) <= 2002:
                    valid = False
                    break
            elif req_field == 'iyr':
                if not 2010 <= int(text) <= 2020:
                    valid = False
                    break
            elif req_field == 'eyr':
                if not 2020 <= int(text) <= 2030:
                    valid = False
                    break
            elif req_field == 'hgt':
                num = int(re.match(r'\d+', text).group())
                if re.search(r'cm', text):
                    if not 150 <= num <= 193:
                        valid = False
                        break
                elif num and re.search(r'in', text):
                    if not 59 <= num <= 76:
                        valid = False
                        break
                else:
                    valid = False
                    break
            elif req_field == 'hcl':
                if not re.search(r'#[0-9a-f]{6}', text):
                    valid = False
                    break
            elif req_field == 'ecl':
                if text not in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']:
                    valid = False
                    break
            elif req_field == 'pid':
                if not len(text) == 9 or not re.match(r'\d{9}', text):
                    valid = False
                    break

        if valid:
            valid_passport_count_part2 += 1


print(f'Part 1: {valid_passport_count}')
print(f'Part 2: {valid_passport_count_part2}')
